### PR TITLE
154 explainability functionality for path based methods and perceived explanation quality metrics

### PR DESCRIPTION
**Pull request by alessandrosocc**
* Fixed an issue where the CKG did not include the user-item (u-i) relation.
* Resolved a problem where the HuggingFace tokenizer mishandled empty strings as input, resulting in paths like [PAD] [PAD] [PAD]. Paths are now filtered to prevent this behavior.
* Added a new sampling method: for each u-i interaction in the training set, k paths from the user to the positive item i are sampled, resulting in an input dataset faithful to the original.
* Set joblib parallelism to use processes instead of threads.
* Added a feature to save multiple dataloaders for models in a subfolder under saved/, organized by model name and dataset. Dataloaders are now automatically retrieved when a configuration matching a previously saved dataloader is set.
* Added Perceived Explanation Quality Metrics (PEQM): 'LIR', 'Fidelity', 'SEP', 'LID', 'SED', 'PTD', 'PTC', 'LITD', 'SETD', 'PPT'. check #153 
* Added support for PEQM evaluation for CAFE, PEARLM, KGGLM, and PGPR.
* Enabled evaluation with different types of rankers and logits processors for PEARLM/KGGLM. See the definitions of Beam Search, Top-k, Top-p, and greedy sampling for further details:
* New Rankers with the 'ranker' parameter:
  * SampleSearchSequenceScoreRanker -> necessary to use topk, topp and greedy sampling
  * CumulativeSequenceScoreRanker -> the score is given by an aggregation of the predicted tokens scores.
  * BeamSearchSequenceScoreRanker -> the score is given by the provided sequence score during beam search
* logits_processor: ConstrainedLogitsProcessorWordLevel (options: 'ConstrainedBeamLogitsProcessorWordLevel', 'ConstrainedLogitsProcessorWordLevel')
* Added SimCE Loss (arXiv:2406.16170).
* Fixed a bug that prevented correct evaluation of some metrics due to a missing collector.
* Fixed torch deprecation warning for weights_only in torch.load.
* Fixed NumPy deprecation warning for np.float.
* Fixed a typo.
* Added functionality to display the Top-k size during evaluation for PEARLM/KGGLM.
* Added the ability to set a process title for better visualization in nvitop, htop, and top.
* Added support for overwriting the checkpoint’s saved configuration. This allows certain configuration parameters to be modified after loading the checkpoint, which would otherwise have been overwritten by the checkpoint's saved config.
* Fixed a bug in resume_checkpoint for Hugging Face-based trainers that caused only the model weights to be loaded, instead of the custom architecture class. This issue led to missing custom methods and embeddings defined in the user-implemented class.
* Fixed the visualization of performance results for the validation and test sets.
* Added pearlm written from scratch with GPT2, LLAMA2 and LLAMA3.1/3.2
* **Added TPRec**
* Added the support for weights training on bfloat16 and float16 precision
* Now the dataloader loaded configuration is overwritten by the configuration the user has set

### CODE REVIEW

**Review by jackmedda — CHANGES_REQUESTED** [review]

File: hopwise/evaluator/collector.py [hunk 1]
```diff
@@ -98,6 +102,57 @@ def train_data_collect(self, train_data):
...
+
+            self.data_struct.set("data.max_path_type", np.arange(train_data.dataset.relation_num))
+        if self.register.need("data.node_degree"):
+            self.data_struct.set("data.node_degree", self.node_degree_dict(train_data))
+        if self.register.need("data.max_path_length"):
+            # how many path constraint are set?
+            # 'path_constraint' is hardcoded and may change in the future
+            if self.config["path_constraint"] is None:
+                # PEARLM or KGGLM
+                sampled_path_len = len(train_data.dataset._path_dataset.split("\n")[0].split(" "))
```
[the remark this answers is not in the document]
Comment (jackmedda) [reply hopwise/evaluator/collector.py]: `train_data_collect` will be called after the mapping and filter. The tokenization process is executed as soon as train_data.tokenized_dataset is accessed, which happens inside hf_path_trainer and before `train_data_collect`

File: hopwise/quick_start/quick_start.py [hunk 2]
```diff
@@ -140,7 +151,9 @@ def run_hopwise(
 
         # model loading and initialization
         init_seed(config["seed"] + config["local_rank"], config["reproducibility"])
-        model = get_model(config["model"])(config, train_data.dataset).to(config["device"])
+        model = get_model(config["model"])(config, train_data.dataset).to(
+            device=config["device"], dtype=config["dtype"]
+        )
```
Comment (jackmedda) [inline hopwise/quick_start/quick_start.py:156]: I have just seen where "dtype" is used and why the configurator has the corresponding function.
I am not sure what happens when this is applied to an entire model. I would say that generally this is fine given that this should not update models' tensors, but only its parameters, which are generally floating points to be used during learning.
I think that HuggingFace does the same, so it's fine. I'm thinking whether this might cause issues if a weight is frozen and converted to integers, but if the tests run smoothly and all models work, then it's fine, as we could provide bfloat16 as supported

Do we have a test to ensure that passing dtype = bfloat16 work as expected? I assume some models might not work because their CUDA operations do not support bfloat16

File: hopwise/quick_start/quick_start.py [hunk 3]
```diff
@@ -151,37 +164,49 @@ def run_hopwise(
...
             train_data, valid_data, saved=saved, show_progress=config["show_progress"]
         )
     elif run == "evaluate":
         if checkpoint is None:
             raise ValueError("Checkpoint is needed for evaluation")
+        trainer.eval_collector.train_data_collect(train_data)
+
+        if isinstance(trainer, HFPathLanguageModelingTrainer):
+            trainer.init_hf_trainer(train_data, valid_data, show_progress=config["show_progress"])
+            trainer.resume_checkpoint(checkpoint, train_data)
```
Comment (jackmedda) [inline hopwise/quick_start/quick_start.py:178]: Indeed the trainer already has train_data anyway, but I think it shouldn't be completely used inside resume_checkpoint

File: hopwise/quick_start/quick_start.py [hunk 4]
```diff
@@ -151,37 +164,49 @@ def run_hopwise(
...
+            for task, result in best_valid_result.items():
+                logger.info(set_color(f"[{task}] best valid ", "yellow") + f": {format_metrics(result)}")
+        else:
+            logger.info(set_color("best valid result", "yellow") + f": {format_metrics(best_valid_result)}")
+
     # model evaluation
     test_result = trainer.evaluate(
-        test_data, load_best_model=True, model_file=checkpoint, show_progress=config["show_progress"]
+        test_data,
+        load_best_model=not isinstance(trainer, HFPathLanguageModelingTrainer),
```
Comment (jackmedda) [inline hopwise/quick_start/quick_start.py:202]: Why this check? The load_best_model param should already be handled inside the HFPathLanguageModelingTrainer as it triggers the internal HuggingFace loading of the best model.

File: hopwise/trainer/hf_path_trainer.py [hunk 5]
```diff
@@ -62,12 +73,17 @@ def reset_topks(self):
 
 
 class CumulativeSequenceScoreRanker:
+    """
+    Ranker that uses the cumulative sequence score of the final 5 predicted tokens to rank sequences.
+    """
+
```
Comment (jackmedda) [inline hopwise/trainer/hf_path_trainer.py:79]: I would suggest having a new file inside hopwise/model named rerankers.py next to layers,py and loss.py and move these rankers there so that this script is neater.
I wouldn't put them inside layers.py as they are not real components of the model, but just processors of their outputs

File: hopwise/trainer/hf_path_trainer.py [hunk 6]
```diff
@@ -168,63 +330,79 @@ def __init__(
...
+                topk=10,
+                max_new_tokens=ranker_max_new_tokens,
+            )
+        else:
+            raise ValueError(
+                f"Ranker {self.hopwise_config['ranker']} not supported. "
+                "Supported rankers are: CumulativeSequenceScoreRanker, "
+                "                       BeamSearchSequenceScoreRanker, "
+                "                       SampleSearchSequenceScoreRanker."
+            )
```
Comment (jackmedda) [inline hopwise/trainer/hf_path_trainer.py:378]: This could be dynamic. Let's create a get_reranker function inside utils, such as get_trainer. Let's use importlib and dynamically pass the parameters instead of this long if-else. Shouldn't topk come from the config?

File: hopwise/config/configurator.py [hunk 7]
```diff
@@ -610,6 +613,25 @@ def _set_eval_neg_sample_args(self, phase: Literal["valid", "test"]):
...
+            dtype = torch.float32
+        elif dtype == "float16":
+            dtype = torch.float16
+        elif dtype == "bfloat16":
+            dtype = torch.bfloat16
+        else:
+            raise ValueError(f"Unsupported dtype: {dtype}")
+
+        self.final_config_dict["dtype"] = dtype
+
```
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/config/configurator.py:634]: I'd consider leaving as I wrote because even though there are many torch dtypes, it makes sense to use only float32, float16 or bfloat16

File: hopwise/model/knowledge_aware_recommender/tprec.py
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/model/knowledge_aware_recommender/tprec.py:1]: we have some changes in get_actions, get_reward fn and in the collect_score fn during evaluation . I'd leave as two separate models considering the confusion that we may create unifying the two models. Also, TPRec have inside TransE to train considering the time aspect of TPRec.

File: hopwise/properties/model/CAFE.yaml [hunk 8]
```diff
@@ -9,9 +9,15 @@ topk_paths: 25
 clip_grad_norm:
   max_norm: 5.0
 
+save_dataset: True
+save_dataloaders: True
+epochs: 50
+eval_step: 50
+topk: [10]
+valid_metric: 'ndcg@10'
+
```
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/properties/model/CAFE.yaml:18]: Actually, in the official code, both PGPR and CAFE train for 50 epochs without tracking metrics

File: hopwise/quick_start/quick_start.py (hunk 2 above)
Comment (alessandrosocc) [reply hopwise/quick_start/quick_start.py:156]: I tested it with PEARLMfromscratch and everything worked fine. It's likely that every torch method support every torch dtype because it's a constraint on the precision of the weights not on other processing phases(torch layers) in how the model makes recommendations... I'd leave as it is, and if there are problems with any model, then it's necessary to use full precision..

File: hopwise/trainer/hf_path_trainer.py (hunk 5 above)
Comment (alessandrosocc) [reply hopwise/trainer/hf_path_trainer.py:79]: I totally agree

File: hopwise/utils/enum_type.py [hunk 9]
```diff
@@ -81,6 +86,7 @@ class InputType(Enum):
     PAIRWISE = 2
     LISTWISE = 3
```
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/utils/enum_type.py:87]: I don't find any usage, but I'd leave it for future use
